API Integration for Europeana

allen505 opened the pull request:
## Fixes
<!-- If PR doesn't fully resolve the issue, replace 'Fixes' below with 'Related to'. -->
<!-- If there is no issue being resolved, open one before creating this pull request. -->
Fixes #279 by @annatuma 

## Description
This PR adds the script `europeana.py` which fetches images from the Europeana Collections and adds them to the cc-catalog. It also adds a test suite under `test_europeana.py` which tests the function in the script.

## Technical details
<!-- Add any other information or technical details about the implementation; or delete the section entirely. -->

- Images are fetched using DelayedRequester class
- Images are stored using ImageStore class
- 100 images can be pulled per request.
- Europeana has around 30 million cc licensed images. Of these, around 20.5 Million images which has direct link to the images.

## Tests
<!-- Give steps for the reviewer to verify that this PR fixes the problem; or delete the section entirely. -->
- To run the script use `python europeana.py`. Optionally `--date [DATE]` can be provided to fetch images created in the specified date.
- To run tests corresponding to the script run `pytest test_europeana.py`

## Checklist
<!-- Replace  the [ ] with [x] to check the boxes. --> 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `master` branch of the repository.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin
<!-- You must read and understand the following attestation. -->

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

allen505 committed:
Added code to fetch details from the first page using Cursor based
pagination.

allen505 committed:
Pagewise function implemented
Images are retrieved pagewise till all images are retrieved

modified:   src/cc_catalog_airflow/dags/provider_api_scripts/europeana.py

allen505 committed:
Stored image & metadata using the ImageStore class

allen505 committed:
Added code to get number of images stored
	modified:   src/cc_catalog_airflow/dags/provider_api_scripts/europeana.py

allen505 committed:
Added test file for Europeana API

Added a new test file for europeana api to run tests
Minor bug fixes in europeana API script

	modified:   dags/provider_api_scripts/europeana.py
	new file:   dags/provider_api_scripts/test_europeana.py

allen505 committed:
Fixed timestamp and cusor bug
Changed the Timestamp to ISO 8601 format
Added code that caused error with last page of the results.

	modified:   dags/provider_api_scripts/europeana.py

allen505 committed:
Added new test functions and response.json file
Test function for empty list and error in response. Minor bugs fixes.
This commit also adds a sample success response to be used for testing
 Changes to be committed:
	modified:   dags/provider_api_scripts/europeana.py
	modified:   dags/provider_api_scripts/test_europeana.py
	new file:   dags/provider_api_scripts/tests/resources/europeana/europeana_example.json

allen505 committed:
LangAware Description, tests to extract_data
Description of images is taken in the following priority:
-English
-Default
-dcDesption if neither of the above were available

New test functions to test extraction of various metadata from the
response
 Changes to be committed:
	modified:   dags/provider_api_scripts/europeana.py
	modified:   dags/provider_api_scripts/test_europeana.py
	modified:   dags/provider_api_scripts/tests/resources/europeana/europeana_example.json
	new file:   dags/provider_api_scripts/tests/resources/europeana/europeana_image_list.json
	new file:   dags/provider_api_scripts/tests/resources/europeana/image_data_example.json

allen505 committed:
Refactor and test function for metadata functions
Test functions for metadata and description functions
Code Reafactoring
New test for get_image_list for last page

 Changes to be committed:
	modified:   europeana.py
	modified:   test_europeana.py

allen505 committed:
Added DAG and corresponding test file
New file to create DAG to execute Europeana's script
Test suite to check for any import errors and the number of DAGs created
 Changes to be committed:
	new file:   dags/europeana_workflow.py
	new file:   dags/test_europeana_workflow.py

mathemancer commented on src/cc_catalog_airflow/dags/europeana_workflow.py:
```diff
@@ -0,0 +1,51 @@
+        dag_id=DAG_ID,
+        default_args=DAG_DEFAULT_ARGS,
+        start_date=datetime(1970, 1, 1),
+        schedule_interval='@monthly',
```
This should be daily, no?
```suggestion
        schedule_interval='@daily',
```

allen505 replied:
I believe we settled on fetching the database monthly as per our conversation on Slack and [this comment](https://github.com/creativecommons/cccatalog/issues/279#issuecomment-611077266).

allen505 replied:
@mathemancer  I concur with your preference. The ingestion method seems to be the most appropriate in our case. 
However, I'm unable to work on this at the moment, so you could take it up.
Thank you!

mathemancer commented on src/cc_catalog_airflow/dags/provider_api_scripts/europeana.py:
```diff
@@ -0,0 +1,282 @@
+
+        logger.debug('response.status_code: {response.status_code}')
+        response_json = _extract_response_json(response)
+        image_list, next_cursor, total_number_of_images = _extract_image_list_from_json(
+            response_json)
```
```suggestion
        (
            image_list,
            next_cursor,
            total_number_of_images
        ) = _extract_image_list_from_json(response_json)
```
Please keep lines to 79 characters or less.

mathemancer commented on src/cc_catalog_airflow/dags/provider_api_scripts/europeana.py:
```diff
@@ -0,0 +1,282 @@
+
+
+def _get_description(image_data):
+    if image_data.get('dcDescriptionLangAware') is not None and image_data.get('dcDescriptionLangAware').get('en') is not None:
+        description = image_data.get('dcDescriptionLangAware').get('en')[0]
+    elif image_data.get('dcDescriptionLangAware') is not None and image_data.get('dcDescriptionLangAware').get('def') is not None:
+        description = image_data.get('dcDescriptionLangAware').get('def')[0]
```
Please use parentheses to make these lines shorter (79 chars or fewer)

mathemancer commented on src/cc_catalog_airflow/dags/provider_api_scripts/europeana.py:
```diff
@@ -0,0 +1,282 @@
+    if len(license_field) > 1:
+        logger.warning('More than one license field found')
+    for license in license_field:
+        cc_license_search = re.search("creativecommons", license)
+        if cc_license_search is not None:
+            return license
```
This isn't complex enough to warrant the import of `re` .  Try this instead:
```suggestion
        if 'creativecommons' in license:
            return license
```

mathemancer commented on src/cc_catalog_airflow/dags/provider_api_scripts/europeana.py:
```diff
@@ -0,0 +1,282 @@
+def _process_image_data(image_data):
+    logger.debug(f'Processing image data: {image_data}')
+    license_url = _get_license_url(image_data.get('rights'))
+    if license_url is None:
+        return None
```
This isn't necessary, since the `image_store.add_item` method will already handle the case where there's no (valid) license URL.

mathemancer reviewed (changes requested):
Fantastic work so far!  I made some specific notes in comments.

Generally, please double-check the line lengths.  They should be limited to 79 characters.

Thank you!

allen505 committed:
Suggestions from code review

Couple of trimming of lines to limit to 79 chars per line
Avoiding the use of RE

Co-authored-by: Brent Moran <brent@creativecommons.org>

allen505 committed:
Removed unnecessary conditions  as per code review
Removed empty license condtions
Trimmed lines to 79 chars per line
Removed import of re

Co-authored-by: mathemancer

allen505 committed:
Trimmed a line europeana.py to fit 79 chars
 Changes to be committed:
	modified:   dags/provider_api_scripts/europeana.py

mathemancer reviewed (changes requested):
I noted some style changes in the main script.

As for the main issue, please see the comment on the workflow.  Basically, we have to either pull all data whenever the script is run (so running it monthly will keep the data fresh), or we need to use the reingestion strategy as per `flickr_ingestion_workflow.py` or `wikimedia_ingestion_workflow.py` in order to keep the data fresh (and then we'd run the workflow daily).

Please let me know if this problem is unclear.

Alternatively:  just make the current workflow daily, and I'll handle setting up the date-partitioned reingestion strategy.

allen505 committed:
Refactored code as per Code review

Changes to some if statement structures to follow PEP8
Changed the dict update parameters

Co-authored-by: Brent Moran <brent@creativecommons.org>

allen505 committed:
Removed return from pagewise and refactored code
Changes were made as per Code Review
Minor changes to fit PEP8
 Changes to be committed:
	modified:   dags/provider_api_scripts/europeana.py

Co-authored-by: Brent Moran <brent@creativecommons.org>

allen505 committed:
Change schedule_interval to daily

Co-authored-by: Brent Moran <brent@creativecommons.org>

allen505 commented:
@mathemancer ,is this pull request considered as a _[Great Contribution](https://opensource.creativecommons.org/community/community-team/project-roles/#great-contributions)_ ?

mathemancer commented on src/cc_catalog_airflow/dags/provider_api_scripts/europeana.py:
```diff
@@ -0,0 +1,285 @@
+        if image_list is not None:
+            break
+
+    if (
+        try_number == max_tries - 1
+        and (image_list is None or next_cursor is None)
+        ):
```
As per PEP8:
```suggestion
    if (
            try_number == max_tries - 1
            and (image_list is None or next_cursor is None)
    ):
```

mathemancer commented on src/cc_catalog_airflow/dags/provider_api_scripts/europeana.py:
```diff
@@ -0,0 +1,285 @@
+        query=f'timestamp_created:[{start_timestamp} TO {end_timestamp}]',
+        cursor=cursor,
+    )
+    
+
```
Please remove the extra line, and make sure blank lines don't contain whitespace characters.
```suggestion

```

mathemancer reviewed (changes requested):
I made a couple of stylistic notes.

The only substantive change I'd ask for is to please add 
```
EUROPEANA_API_KEY=not_set
```
to the `src/cc_catalog_airflow/env.template` file.  Please use alphabetical order, so it'd look like:
```
DATA_GOV_API_KEY=not_set
EUROPEANA_API_KEY=not_set
FLICKR_API_KEY=not_set
```
Sorry I forgot about that necessity earlier.

Thank you!

allen505 committed:
Code refactoring as per Code review

Co-authored-by: Brent Moran <brent@creativecommons.org>

allen505 committed:
Added european key to env.template
 Changes to be committed:
	modified:   env.template

Co-authored-by: Brent Moran <brent@creativecommons.org>

allen505 committed:
Merge branch 'master' into europeana

mathemancer reviewed (approved):
This looks great, thank you so much for your contribution!
